store/driver/txn/ranged_kv_retriever.go

```diff
@@ -62,3 +65,158 @@ func (s *RangedKVRetriever) Intersect(startKey, endKey kv.Key) *RangedKVRetrieve
 
 	return nil
 }
+
+// ScanCurrentRange scans the retriever for current range
+func (s *RangedKVRetriever) ScanCurrentRange(reverse bool) (kv.Iterator, error) {
+	if !s.Valid() {
+		return nil, errors.New("retriever is invalid")
+	}
+
+	startKey, endKey := s.StartKey, s.EndKey
+	if !reverse {
+		return s.Iter(startKey, endKey)
+	}
+
+	iter, err := s.IterReverse(endKey)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(startKey) > 0 {
+		iter = newLowerBoundReverseIter(iter, startKey)
+	}
+
+	return iter, nil
+}
+
+type sortedRetrievers []*RangedKVRetriever
+
+func (retrievers sortedRetrievers) TryGet(ctx context.Context, k kv.Key) (bool, []byte, error) {
+	if len(retrievers) == 0 {
+		return false, nil, nil
+	}
+
+	_, r := retrievers.searchRetrieverByKey(k)
+	if r == nil || !r.Contains(k) {
+		return false, nil, nil
+	}
+
+	val, err := r.Get(ctx, k)
+	return true, val, err
+}
+
+func (retrievers sortedRetrievers) TryBatchGet(ctx context.Context, keys []kv.Key, collectF func(k kv.Key, v []byte)) ([]kv.Key, error) {
+	if len(retrievers) == 0 {
+		return keys, nil
+	}
+
+	// assign keys to nonCustomKeys, if all the keys are non-custom, no new space will be allocated
+	nonCustomKeys := keys
+	hasCustomKeys := false
+	for i, k := range keys {
+		custom, val, err := retrievers.TryGet(ctx, k)
+		if !custom {
+			// hasCustomKeys means that nonCustomKeys is a newly allocated slice, so we must append the key to its end.
+			if hasCustomKeys {
+				nonCustomKeys = append(nonCustomKeys, k)
+			}
+			continue
+		}
+
+		if !hasCustomKeys {
+			// If we encounter a custom key first time, we must alloc a new slice and append all the keys before i into it.
+			hasCustomKeys = true
+			nonCustomKeys = append(make([]kv.Key, 0, i), nonCustomKeys[:i]...)
+		}
```

Comment:
This code is a bit tricky.
Are you trying to reuse the keys memory here?

Reply:
Yes, its a optimize

Reply:
Mix the allocation optimization and the function logic make the code hard to read.
It's better to trade performance for readability here.

Reply:
Fine